### DISCUSSION TITLE
How to distribute/consume API + Impl from Maven in Target Platform?

### DISCUSSION THREAD

**Discussion by guw** [eclipse-pde/eclipse.pde#267] [Q&A]
I am looking at consuming artifacts directly from Maven instead of Orbit. The artifacts contain of two elements:

```
		<location includeDependencyDepth="none" includeSource="true" missingManifest="generate" type="Maven">
			<dependencies>
				<dependency>
                    <groupId>com.sun.activation</groupId>
                    <artifactId>jakarta.activation</artifactId>
                    <version>1.2.2</version>
					<type>jar</type>
				</dependency>
				<dependency>
                    <groupId>jakarta.xml.bind</groupId>
                    <artifactId>jakarta.xml.bind-api</artifactId>
                    <version>2.3.3</version>
					<type>jar</type>
				</dependency>
				<dependency>
                    <groupId>com.sun.xml.bind</groupId>
                    <artifactId>jaxb-impl</artifactId>
                    <version>2.3.3</version>
					<type>jar</type>
				</dependency>
			</dependencies>
		</location>
```

Currently in the OSGi manifest of my bundle I import only the package `javax.xml.bind`. 

Anyone has any guidance/tips for how to ensure that the target environment my bundles get installed to will have both - the JAXB api and the implementation available?

Does the Maven support know about those common cases and create proper OSGi headers so the API depends on Impl?
Should I include the generated bundles in my feature to be sure?

**Comment by laeubi** [comment]
> I import only the package `javax.xml.bind`.

Just in case I can only recommend migrating to `jakarta.xml.*` and you most probably want to include the `org.glassfish.hk2.locator`

> Anyone has any guidance/tips for how to ensure that the target environment my bundles get installed to will have both - the JAXB api and the implementation available?

The usual way is to either create a feature with proper requirements or using p2.inf if you are talking about Eclipse/P2 managed environment.

> Does the Maven support know about those common cases and create proper OSGi headers so the API depends on Impl?

By default, only optional requirements are produced, even if they are required the actually only require *any* implementation but not *that* implementation.

**Reply by guw** [reply]
Do you have an example of what artifacts/bundles I will need in combination with `org.glassfish.hk2.locator`?

**Reply by laeubi** [reply]
`org.glassfish.hk2.locator` should work with all the recent jakarta stuff (JAXB, JAX-WS, JAX-RS), just don'T forget to start it in your product because I think it requires the bundle to be active to be picked up by the jakarta stuff.
